File: sdk/typescript/src/networks.ts
import { CommandFailedError } from "./errors.js";
import { runCli } from "./process.js";
import { CommandResult } from "./shell.js";
import { Sandbox } from "./sandbox.js";
import type { NetworkInfo, SandboxInfo } from "./types.js";

function mapNetwork(row: Record<string, unknown>): NetworkInfo {
  return {
    name: String(row.name),
    subnet: String(row.subnet),
    gateway: String(row.gateway),
    members: Number(row.members ?? 0),
    running: Number(row.running ?? 0),
    up: Boolean(row.up),
    createdAt: row.created_at == null ? null : Number(row.created_at),
  };
}

async function run(args: string[], label: string): Promise<CommandResult> {
  const res = await runCli(args);
  const result = new CommandResult(res.stdout, res.stderr, res.exitCode, label);
  if (res.exitCode !== 0) throw new CommandFailedError(result);
  return result;
}

/** List global networks and their member counts. */
export async function listNetworks(): Promise<NetworkInfo[]> {
  const res = await run(["network", "ls", "--json"], "bsdkrun network ls");
  const rows = JSON.parse(res.stdout || "[]") as Record<string, unknown>[];
  return rows.map(mapNetwork);
}

/** Create a global network (starts its shared switch). */
export async function createNetwork(name: string): Promise<void> {
  await run(["network", "create", name], "bsdkrun network create");
}

/** Remove one or more networks. `force` allows removal with running members. */
export async function removeNetwork(
  names: string | string[],
  opts: { force?: boolean } = {},
): Promise<void> {
  const list = Array.isArray(names) ? names : [names];
  const args = ["network", "rm"];
  if (opts.force) args.push("--force");
  args.push(...list);
  await run(args, "bsdkrun network rm");
}

/**
 * Join or switch a machine to a network (by id or name). Applies on the
 * machine's next {@link Sandbox.start}.
 */
export async function connectNetwork(
  machine: string,
  network: string,
): Promise<void> {
  await run(
    ["network", "connect", machine, network],
    "bsdkrun network connect",
  );
}

/** Detach a machine from its network. Applies on its next start. */
export async function disconnectNetwork(machine: string): Promise<void> {
  await run(["network", "disconnect", machine], "bsdkrun network disconnect");
}

/**
 * Refresh members' `/etc/hosts` with the current membership so peers resolve by
 * name (notably NetBSD). Joins auto-sync; use this to fix an existing network
 * without restarting members.
 */
export async function syncNetwork(network: string): Promise<void> {
  await run(["network", "sync", network], "bsdkrun network sync");
}

/** The machines currently attached to `network` (running or stopped). */
export async function networkMembers(network: string): Promise<SandboxInfo[]> {
  const all = await Sandbox.list({ all: true });
  return all.filter((m) => m.network === network);
}

/** Namespace grouping global-network operations. */
export const networks = {
  list: listNetworks,
  create: createNetwork,
  remove: removeNetwork,
  connect: connectNetwork,
  disconnect: disconnectNetwork,
  sync: syncNetwork,
  members: networkMembers,
};

File: sdk/rust/src/networks.rs
//! Global-network operations — reach machines by name on a shared subnet.

use serde_json::Value;

use crate::error::Result;
use crate::process::run_checked;
use crate::sandbox::Sandbox;
use crate::types::{NetworkInfo, SandboxInfo};

/// List global networks and their member counts.
pub fn list() -> Result<Vec<NetworkInfo>> {
    let res = run_checked(["network", "ls", "--json"], "bsdkrun network ls")?;
    let raw = if res.stdout.trim().is_empty() {
        "[]".to_string()
    } else {
        res.stdout
    };
    let rows: Value = serde_json::from_str(&raw)?;
    Ok(rows
        .as_array()
        .map(|rows| rows.iter().map(NetworkInfo::from_row).collect())
        .unwrap_or_default())
}

/// Create a global network (starts its shared switch).
pub fn create(name: &str) -> Result<()> {
    run_checked(["network", "create", name], "bsdkrun network create")?;
    Ok(())
}

/// Remove one or more networks. `force` allows removal with live members.
pub fn remove<S: AsRef<str>>(names: &[S], force: bool) -> Result<()> {
    let mut args = vec!["network".to_string(), "rm".to_string()];
    if force {
        args.push("--force".to_string());
    }
    args.extend(names.iter().map(|n| n.as_ref().to_string()));
    run_checked(args, "bsdkrun network rm")?;
    Ok(())
}

/// Join or switch a machine (by id or name) to a network (next start).
pub fn connect(machine: &str, network: &str) -> Result<()> {
    run_checked(
        ["network", "connect", machine, network],
        "bsdkrun network connect",
    )?;
    Ok(())
}

/// Detach a machine from its network. Applies on its next start.
pub fn disconnect(machine: &str) -> Result<()> {
    run_checked(
        ["network", "disconnect", machine],
        "bsdkrun network disconnect",
    )?;
    Ok(())
}

/// Refresh members' `/etc/hosts` so peers resolve by name (notably NetBSD).
pub fn sync(network: &str) -> Result<()> {
    run_checked(["network", "sync", network], "bsdkrun network sync")?;
    Ok(())
}

/// The machines currently attached to `network` (running or stopped).
pub fn members(network: &str) -> Result<Vec<SandboxInfo>> {
    Ok(Sandbox::list(true)?
        .into_iter()
        .filter(|m| m.network.as_deref() == Some(network))
        .collect())
}
